File: testng-core/src/test/java/test/hook/samples/HookSuccessDynamicParametersSample.java
package test.hook.samples;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import javax.inject.Named;
import org.testng.Assert;
import org.testng.IHookCallBack;
import org.testng.IHookable;
import org.testng.ITestResult;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

public class HookSuccessDynamicParametersSample implements IHookable {

  @Override
  public void run(IHookCallBack callBack, ITestResult testResult) {
    Method method = testResult.getMethod().getConstructorOrMethod().getMethod();
    for (int i = 0; i < callBack.getParameters().length; i++) {
      Annotation[] annotations = method.getParameterAnnotations()[i];
      for (Annotation annotation : annotations) {
        if (annotation instanceof Named) {
          Named named = (Named) annotation;
          callBack.getParameters()[0] = callBack.getParameters()[0] + named.value();
        }
      }
    }
    callBack.runTestMethod(testResult);
  }

  @DataProvider
  public Object[][] dp() {
    return new Object[][] {new Object[] {"foo", "test"}};
  }

  @Test(dataProvider = "dp")
  public void verify(@Named("bar") String bar, String test) {
    Assert.assertEquals(bar, "foobar");
    Assert.assertEquals(test, "test");
  }
}

File: testng-core/src/test/kotlin/test/hook/samples/HookSuccessDynamicParametersSample.kt
package test.hook.samples

import javax.inject.Named
import org.testng.Assert
import org.testng.IHookCallBack
import org.testng.IHookable
import org.testng.ITestResult
import org.testng.annotations.DataProvider
import org.testng.annotations.Test
import test.TestData

class HookSuccessDynamicParametersSample : IHookable {

    override fun run(callBack: IHookCallBack, testResult: ITestResult) {
        val method = testResult.method.constructorOrMethod.method
        for (i in 0 until callBack.parameters.size) {
            method.parameterAnnotations[i].forEach { annotation ->
                annotation.takeIf { it is Named }.run {
                    val named = this as Named
                    callBack.parameters[0] = callBack.parameters[0].toString() + named.value
                }
            }
            callBack.runTestMethod(testResult)
        }
    }

    @DataProvider
    fun dp(): TestData = arrayOf(
        arrayOf("foo", "test")
    )

    @Test(dataProvider = "dp")
    fun verify(@Named("bar") bar: String, test: String) {
        Assert.assertEquals(bar, "foobar")
        Assert.assertEquals(test, "test")
    }
}
